src/main/java/service/OrderService.java

```diff
@@ -0,0 +1,97 @@
+
+import java.util.List;
+
+public class OrderService {
+
+    private final OrderHistoryRepository orderHistoryRepository = new OrderHistoryRepository();
+    private final OrderGroupRepository orderGroupRepository = new OrderGroupRepository();
+    private final ProductRepository productRepository =  new ProductRepository();
+    private final UserRepository userRepository = new UserRepository();
+
+
+    // 주문 기능
+    // 트랜잭션 처리 됐다고 가정 ~~
+    public void orderProducts(List<OrderProductRequestDto> orderInfos, String userName) {
+
+        OrderGroup orderGroup = orderGroupRepository.save(new OrderGroup(orderGroupRepository.getSequence()));
+        User currentUser = userRepository.findByName(userName);
+
+        for (OrderProductRequestDto orderInfo : orderInfos) {
+
+            String productCode = orderInfo.getProductCode();
+            Product product = productRepository.findByCode(productCode);
+
+            int quantity = orderInfo.getQuantity();
+            int price = product.getPrice();
+            int amount = quantity * price;
+
+            OrderHistory orderHistory = new OrderHistory(orderHistoryRepository.getSequence(), orderGroup.getId(), product.getCode(), quantity, price ,Status.PAID, currentUser.getName());
+            orderHistoryRepository.save(orderHistory);
+
+            product.deductStock(quantity);
+            currentUser.chargeBalance(amount);
+        }
+    }
+
+    // 전체 취소
+    // 트랜잭션 처리 됐다고 가정 ~~
+    public void cancelOrders(Long groupId) {
+
+        List<OrderHistory> orders = orderHistoryRepository.findByOrderGroupId(groupId);
+        for (OrderHistory order : orders) {
+            cancelOrder(order);
+        }
+    }
+
+
+    // 부분 취소
+    // 트랜잭션 처리 됐다고 가정 ~~
+    public void cancelOrder(Long orderHistoryId) {
+
+        OrderHistory order = orderHistoryRepository.findById(orderHistoryId);
+        cancelOrder(order);
+    }
+
+    // 상품별 판매 금액 조회
+    public int getSalesByProduct(String productCode) {
+
+        int result = 0;
+
+        List<OrderHistory> orders = orderHistoryRepository.findByProductAndStatus(productCode, Status.PAID);
+
+        for (OrderHistory order : orders) {
+            int price = order.getPrice();
+            int quantity = order.getQuantity();
+
+            result += quantity * price;
+        }
+        return result;
+    }
+
+
+    private void cancelOrder(OrderHistory order) {
+
+        Status status = order.getStatus();
+        if (status == Status.CANCELLED) {throw new IllegalStateException("이미 취소된 주문입니다.");}
+
+        Product product = productRepository.findByCode(order.getProductCode());
+        User user = userRepository.findByName(order.getUserName());
+
+        int orderQuantity = order.getQuantity();
+        int orderPrice = order.getPrice();
+        int amount = orderQuantity * orderPrice;
+
+
+        order.cancelStatus();
+        user.refundBalance(amount);
+        product.refundStock(orderQuantity);
+    }
```

Comment:
서비스로직에 도메인 모델 안으로 들어가도 괜찮을 부분들이 많이 올라와있는 것 같아요.
예를 들면 아래와 같이 개선을 해볼 수 있을 것 같다는 생각이 듭니다

```suggestion
    private void cancelOrder(OrderHistory order) {
        
        // 이런 식으로 도메인 모델에게 직접 취소여부를 물어보는 방식
        if (order.isCanceled()) {throw new IllegalStateException("이미 취소된 주문입니다.");}
        

        Product product = productRepository.findByCode(order.getProductCode());
        User user = userRepository.findByName(order.getUserName());

        // int orderQuantity = order.getQuantity();
        // int orderPrice = order.getPrice();
        // int amount = orderQuantity * orderPrice;

        // 이런식으로 연산에 대한 부분을 도메인 모델 안으로 밀어넣기
        int amount = order.getTotalPrice();

        order.cancelStatus();
        user.refundBalance(amount);
        product.refundStock(orderQuantity);
    }
```

Reply:
저도 동의합니다! 어디까지가 서비스고 어디까지가 도메인 로직으로 들어갈지 고민될 때에는(저도 고민자주했음) - 도메인 객체 자체의 일관성을 보장해야 하는 규칙 (항상 지켜져야 하는 규칙) 이 무엇인지 고민해보면 좋을 것 같아요.

Reply:
공감추